File: example-base-17/backend/routes/manager.js
const express = require('express');
const router = express.Router();
const { authenticate, allowRoles } = require('../middleware/auth');
const User = require('../models/User');

// All manager routes require authentication and MANAGER or ADMIN role
router.use(authenticate);
router.use(allowRoles('ADMIN', 'MANAGER'));

// Get limited users list (Manager can read, but not all details)
router.get('/users', async (req, res) => {
  try {
    // Managers can only see USER role users (not other managers or admins)
    const users = await User.find({ role: 'USER' })
      .select('username email role createdAt')
      .limit(50); // Limit results
    
    res.json({
      message: 'Limited users retrieved successfully',
      count: users.length,
      users
    });
  } catch (error) {
    console.error('Get users error:', error);
    res.status(500).json({ message: 'Server error', error: error.message });
  }
});

// Update user (limited - Manager can only update USER role accounts)
router.patch('/users/:id', async (req, res) => {
  try {
    const user = await User.findById(req.params.id);
    
    if (!user) {
      return res.status(404).json({ message: 'User not found' });
    }

    // Managers cannot update ADMIN or MANAGER accounts
    if (user.role === 'ADMIN' || user.role === 'MANAGER') {
      return res.status(403).json({ 
        message: 'Forbidden: Cannot update admin or manager accounts' 
      });
    }

    // Allow updating username and email (not password or role)
    const { username, email } = req.body;
    if (username) user.username = username;
    if (email) user.email = email;

    await user.save();

    res.json({
      message: 'User updated successfully',
      user: {
        id: user._id,
        username: user.username,
        email: user.email,
        role: user.role
      }
    });
  } catch (error) {
    console.error('Update user error:', error);
    res.status(500).json({ message: 'Server error', error: error.message });
  }
});

// Manager dashboard
router.get('/dashboard', (req, res) => {
  res.json({
    message: 'Welcome to Manager Dashboard',
    user: req.user,
    access: 'Limited resource access - can read/update USER accounts only'
  });
});

module.exports = router;

File: example-base-17/backend-python/app/routers/manager.py
from fastapi import APIRouter, Depends, HTTPException, status
from app.schemas.user import UpdateUserRequest, DashboardResponse
from app.models.user import Role
from app.repositories.user_repository import UserRepository
from app.database import get_database
from app.dependencies import allow_roles
from motor.motor_asyncio import AsyncIOMotorDatabase

router = APIRouter()

@router.get("/users")
async def get_users(
    db: AsyncIOMotorDatabase = Depends(get_database),
    current_user: dict = Depends(allow_roles("ADMIN", "MANAGER"))
):
    user_repo = UserRepository(db)
    # Managers can only see USER role users
    users = await user_repo.find_by_role(Role.USER)
    
    user_responses = []
    for user in users[:50]:  # Limit results
        user_responses.append({
            "id": user["id"],
            "username": user["username"],
            "email": user["email"],
            "role": user["role"],
            "createdAt": user.get("createdAt")
        })
    
    return {
        "message": "Limited users retrieved successfully",
        "count": len(user_responses),
        "users": user_responses
    }

@router.patch("/users/{user_id}")
async def update_user(
    user_id: str,
    request: UpdateUserRequest,
    db: AsyncIOMotorDatabase = Depends(get_database),
    current_user: dict = Depends(allow_roles("ADMIN", "MANAGER"))
):
    user_repo = UserRepository(db)
    user = await user_repo.find_by_id(user_id)
    
    if not user:
        raise HTTPException(
            status_code=status.HTTP_404_NOT_FOUND,
            detail="User not found"
        )
    
    # Managers cannot update ADMIN or MANAGER accounts
    if user["role"] in ["ADMIN", "MANAGER"]:
        raise HTTPException(
            status_code=status.HTTP_403_FORBIDDEN,
            detail="Forbidden: Cannot update admin or manager accounts"
        )
    
    updated_user = await user_repo.update_user(
        user_id,
        username=request.username,
        email=request.email
    )
    
    if not updated_user:
        raise HTTPException(
            status_code=status.HTTP_500_INTERNAL_SERVER_ERROR,
            detail="Failed to update user"
        )
    
    return {
        "message": "User updated successfully",
        "user": {
            "id": updated_user["id"],
            "username": updated_user["username"],
            "email": updated_user["email"],
            "role": updated_user["role"]
        }
    }

@router.get("/dashboard", response_model=DashboardResponse)
async def get_dashboard(current_user: dict = Depends(allow_roles("ADMIN", "MANAGER"))):
    return DashboardResponse(
        message="Welcome to Manager Dashboard",
        username=current_user["username"],
        role=current_user["role"],
        access="Limited resource access - can read/update USER accounts only"
    )

